File: MMCardView/Classes/_DelegateProxy.m
//
//  _DelegateProxy.m
//  Pods
//
//  Created by Millman YANG on 2017/6/21.
//
//

#import "_DelegateProxy.h"

BOOL method_signature_void(NSMethodSignature * __nonnull methodSignature) {
    const char *methodReturnType = methodSignature.methodReturnType;
    return strcmp(methodReturnType, @encode(void)) == 0;
}


@interface _DelegateProxy () {
    id __weak __forwardToDelegate;
}
@end

@implementation _DelegateProxy

-(id)_forwardToDelegate {
    return __forwardToDelegate;
}

-(BOOL)respondsToSelector:(SEL)aSelector {
    
    return [self._forwardToDelegate respondsToSelector:aSelector] || [super respondsToSelector:aSelector];
}

-(void)setForward:(id __nullable)forwardDelegate {
    __forwardToDelegate = forwardDelegate;
}

- (void)forwardInvocation:(NSInvocation *)invocation {
    if (self.parent && [self.parent respondsToSelector:invocation.selector]) {
        [invocation invokeWithTarget:self.parent];
    } else if (self._forwardToDelegate && [self._forwardToDelegate respondsToSelector:invocation.selector]) {
        [invocation invokeWithTarget:self._forwardToDelegate];
    }
}

-(void)_methodInvoked:(SEL)selector {
    
}
@end

File: MMCardView/Classes/DelegateProxy.swift
//
//  DelegateProxy.swift
//  Pods
//
//  Created by Millman YANG on 2017/6/21.
//
//

import UIKit

class DelegateProxy: NSObject ,UICollectionViewDelegateFlowLayout {
    unowned let parent: AnyObject
    public weak var forwardDelegate: AnyObject?
    public init(parentObject: AnyObject) {
        self.parent = parentObject
        super.init()
    }
    
    override func forwardingTarget(for aSelector: Selector!) -> Any? {
        if self.parent.responds(to: aSelector) {
            return parent
        } else if let forward = self.forwardDelegate, forward.responds(to: aSelector){
            return forward
        }
        return super.forwardingTarget(for: aSelector)
    }
    
    override func responds(to aSelector: Selector!) -> Bool {
        if self.parent.responds(to: aSelector) {
            return true
        } else if let forward = self.forwardDelegate {
            return forward.responds(to: aSelector)
        }
        return super.responds(to: aSelector)
    }
}
